### PR TITLE
feat(invariant): introduce `afterInvariant` function

### LINKED ISSUE #4300: feat: Add "after-all" hook for testing
### Component

Forge

### Describe the feature you would like

In `forge` testing it would be very helpful to have a hook that runs at the end of each test. This would allow for:
- Single logging of metrics as outlined in this [issue](https://github.com/foundry-rs/foundry/issues/3607) as an intermediate measure until native internal metrics are implemented.
- Post fuzz campaign testing (e.g., close out all positions and assert all funds are able to exit the system) (both fuzzing and invariant testing).

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
Closes #4300 - This PR adds `afterInvariant` for invariant tests.
 

## Solution

- add `afterInvariant` hook, test fails if multiple functions defined, warning is raised if wrong sig
- called after each successful invariant run when shrinking/replaying failure and when traces collected
- code cleanup: add `call_invariant` and `call_after_invariant` helpers
- added tests to cover
  -  invariant failure, afterInvariant is not called: test fails
  - invariant pass, failure in afterInvariant: test fails
  - both invariant and afterInvariant succeed: test pass

### CODE REVIEW

File: crates/evm/evm/src/executors/invariant/result.rs [hunk 1]
```diff
@@ -155,3 +147,45 @@ pub(crate) fn can_continue(
...
+            call_result,
+            &[],
+        );
+        invariant_failures.error = Some(InvariantFuzzError::BrokenInvariant(case_data));
+    }
+    Ok(success)
+}
+
+/// Calls the invariant function and returns call result and if succeeded.
+pub(crate) fn call_invariant_function(
```
Comment (mattsse) [inline crates/evm/evm/src/executors/invariant/result.rs:178]: should we move this to executor: `fn (self: &Executor)` ?

File: crates/evm/evm/src/executors/invariant/result.rs [hunk 2]
```diff
@@ -155,3 +147,45 @@ pub(crate) fn can_continue(
...
     Ok(RichInvariantResults::new(true, call_results))
 }
+
+/// Given the executor state, asserts conditions within `tearDown` function.
+/// If call fails then the invariant test is considered failed.
+pub(crate) fn assert_tear_down(
+    invariant_contract: &InvariantContract<'_>,
+    invariant_config: &InvariantConfig,
+    targeted_contracts: &FuzzRunIdentifiedContracts,
+    executor: &mut Executor,
```
Comment (mattsse) [inline crates/evm/evm/src/executors/invariant/result.rs:157]: same q,

but perhaps we keep it in variant because these functions are only useful in here?

**Review by mattsse — COMMENTED** [review]
I think this will work.

do we want to limit teardown to invariant runs or should teardown general feature?

**Comment by grandizzy** [comment]
> I think this will work.
> 
> do we want to limit teardown to invariant runs or should teardown general feature?

Ideally it should be a general feature, but for regular and fuzz tests it's a little bit more complex to add (as listed above they don't commit state, should also work for debug and should respect the failure / replay persistence for fuzz tests).
The simplest way I can think of is to make a multi call test fn + teardown if tearDown defined but needs forge-std support too - suggestions on how to better tackle this appreciated :pray:

File: crates/evm/evm/src/executors/invariant/result.rs (hunk 2 above)
Comment (grandizzy) [reply crates/evm/evm/src/executors/invariant/result.rs:157]: Yeah, they're invariant related only and cannot be reused for other types of test, so probably better to keep them here

**Comment by mattsse** [comment]
> The simplest way I can think of is to make a multi call test fn + teardown if tearDown 

ah I see, but effectively you'd get the same if you'd call some teardown function manually?

re
https://github.com/foundry-rs/foundry/issues/4300#issuecomment-1447463993

perhaps we need to come up with use case specific naming?

wdyt @DaniPopes @mds1 @klkvr

**Comment by grandizzy** [comment]
discussed internally - will be supporting this feature for invariant and renaming `tearDown` -> `afterInvariant`, for regular and fuzz tests this can be easily done by directly calling function from tests (or by using a modifier)

File: crates/forge/src/runner.rs [hunk 3]
```diff
@@ -310,6 +309,30 @@ impl<'a> ContractRunner<'a> {
         }
 
         let has_invariants = self.contract.abi.functions().any(|func| func.is_invariant_test());
```
Comment (mattsse) [inline crates/forge/src/runner.rs:311]: this call looks out of place here,
can we move this closer to where we check this?

**Review by mattsse — APPROVED** [review]
okay, this is effectively, post-processing of invariant tests, invoked.

the afterInvariant is defined on the contract level, and invoked on all tests, same as setup
I think this is reasonable. if a user needs different afterInvariant handlers, they should be moved to separate contracts.

pending @DaniPopes @klkvr

File: crates/forge/src/runner.rs (hunk 3 above)
Comment (grandizzy) [reply crates/forge/src/runner.rs]: yep, grouped it with tmp_tracing in https://github.com/foundry-rs/foundry/pull/8106/commits/21a790b52180ddc0d96a8c3cf89a880f5c8ca50e

**Review by klkvr — APPROVED** [review]
lgtm!

**Pull request merged**